### DISCUSSION TITLE
How should assigning Xpress proposals to techniques and instruments work?

### DISCUSSION THREAD

**Discussion by simonfernandes** [UserOfficeProject/user-office-project-issue-tracker#1032] [General]
This discussion should be used to track the design decisions we make for the Xpress technique/instrument relationship, the final decisions should be tracked in this top comment.

---

**Question** 

We will be asking users to submit their Xpress proposals for a particular technique and not instrument. How do we go about this? 

Scientists need to receive notifications and have management controls for proposals within their technique. We already maintain a list of instruments and their scientists in the system so should we somehow map techniques to instruments?

**Decisions**
- Create "techniques" that can have instruments attached, via a separate area/tab.
  - Attach people to the call for who should receive the initial notification and potentially for some permission configuration (TBC see discussion in https://github.com/UserOfficeProject/issue-tracker/issues/1079)
- Assign instruments to the call as we would normally do when using the instrument picker. We wouldn't need to assign techniques.
- User selects a technique in questionnaire via a new technique picker question.
- The technique picker question would display (unique) techniques based on the instruments attached to the call.
- In the new management page, only allow scientists to pick instruments that the technique is associated with. We can also filter out decommissioned instruments that are in the technique but weren't added to the call.
- The instrument is assigned to the proposal in the management page without the instrument picker.
- An Xpress proposal is required to have an instrument assigned to it.

**Previous suggestions:**

1. Have a separate call for each technique, and assign one of the technique's instruments to the call.
    - Would require some kind of new interface for picking Xpress proposals, to avoid loads of calls in a single list.
    - Flexibility to use a common questionnaire or use separate but similar ones.
2. Use the instrument picker question type with techniques.
    - We would need to add techniques as instruments and maintain the list of scientists associated with it
    - How is an instrument chosen?
3. Assign techniques to instruments.

**Comment by simonfernandes** [comment]
> 1. Have a separate call for each technique, and assign one of the technique's instruments to the call.

My thoughts:

1. The main complexity seems like it would be finding a way to display all of the Xpress calls in a nice interface, so that we don't end up with a huge list of calls.
2. Down the line, would we also want a way to bulk manage these Xpress calls? E.g. to open all, close all, etc. With the new email system being configured in the call workflow, and with the amount of emails sent from Xpress, there will be a lot of setup per call.
3. An easier solution might be to have a single call (behind the scenes) that can be created x times with a pre-selected instrument, if techniques are happy to share the many options and use a single questionnaire.
4. The scientists assigned to instruments in Submissions differ a lot from the admins currently in legacy Xpress. E.g. for Chronus, there are 10 scientists in Submissions, 3 of which are admins in legacy Xpress out of 4 total. If not all Submissions scientists should have access or receive notifications, will we need separate Xpress instruments in Submissions?

**Comment by simonfernandes** [comment]
> 2. Use the instrument picker question type with techniques.

My thoughts:

1. It would mean creating techniques as experimental areas and using it in the instrument picker.
2. For assigning an instrument to a technique: each experimental area currently has a "name" and a "short code". Is it possible for us to use one field to present the option as a technique to the user, but call it an instrument in the other, so that allocations ingests it as such? At this point it might just be the inverse of assigning techniques to instruments, but with less work needed perhaps.
3. If a technique ever needed to be assigned to a different instrument, we might have to create a new technique to avoid breaking existing proposlas, which could get messy.

**Comment by simonfernandes** [comment]
> 3. Assign techniques to instruments.

My thoughts:

1. The concept of applying techniques to instruments sounds most logical and nicest on paper. Is there possible future benefit to this outside of Xpress?
2. If we use this in conjunction with the instrument picker, I think we'd still need a way to present the instruments as techniques to the user.
3. If we need separate instruments for Xpress, then this might have less value.

**Comment by simonfernandes** [comment]
Discussion with Farai 30th May 2024:

Proposed solution:

- Create "techniques" that can have instruments attached, via a separate area/tab.
  - We may need to create new Xpress instruments if the list of Xpress scientists differs to normal. In that case we need to check whether the instrument can be named the same without conflict.
  - Leave the regular (non Xpress) instruments without a technique, only assign the Xpress ones.
  - The technique could optionally include configuration for permissions for who should be able to assign instrument
  - The technique could optionally include configuration for who should receive the emails, if that ever needs to be configured on a per-technique basis.
- Assign instruments to the call as we would normally do when using the instrument picker. We wouldn't need to assign techniques.
- User selects a technique in questionnaire via a new technique picker question.
  - The technique picker question would display (unique) techniques based on the instruments attached to the call.
- In the new management page, only allow scientists to pick instruments that the technique is associated with. We can also filter out decommissioned instruments that are in the technique but weren't added to the call.
- The instrument is assigned to the proposal in the management page without the instrument picker.

Other thoughts:
- An option in the call might be the best way to indicate when proposals in a call should be managed by the new instrument management page.
- If a user needs to know if the proposal was successful, we could either include it in the success email, or we may need to show Xpress proposal statuses on the user dashboard. Opened https://github.com/UserOfficeProject/issue-tracker/issues/1090

**Reply by simonfernandes** [reply]
Notes from design review:

- We shouldn't need separate Xpress instruments, but if do we need to make sure downstream software is configured to look at them. Or we can perhaps use the same short code and a different name for each instrument.
- Ideally for the first phase, the technique should have two lists of names attached to it: for 1) for who should receive the initial notification for the technique and 2) who has permission to assign an instrument.
- Worth talking to R&A about the option in the call, as they are doing similar for technical reviews.
- We do need to show users the status of their proposal within the software (extra info added to https://github.com/UserOfficeProject/issue-tracker/issues/1090)

**Comment by bolmsten** [comment]
Comments from ESS:

- We agree that adding a new tab with all the techniques and attaching them to instruments are a good idea. 
- We suggest to make parts of the technical and management pages dynamic, to cater for the new needs of more fields. 
- We suggest that the technique picker assigns the proposal to all the instruments that are applicable. 
- We suggest to add a new status action "Remove instruments without technical review" 
- Create new proposal status that allow for instrument scientist to approve proposals called "management decision by instrument scientist"
- Create a new proposal workflow which goes from technical review to "management decision by instrument scientist" that has the status action "Remove instruments without technical review"

**Reply by simonfernandes** [reply]
@bolmsten We discussed this as a team and we like generally think this is a good and consistent approach.

One thing we would like to explore is your 3rd bullet point, and whether it is better and more intuitive to automatically assign 0 instruments and then have a stage to add them, instead of the other way around. Is there a reason you suggested to doing it this way?

**Reply by bolmsten** [reply]
@martin-trajanovski , please give input here as well.

I think the reason was that you wanted all the instrument people that were interested should see the proposal, so by assigning it directly to their instrument you do not need to add new functionality for techniques.

**Reply by martin-trajanovski** [reply]
Yes I agree. That way we will inform and notify all the people that can look into the specific proposal and then the approach is that the one who takes the first review stays on the proposal and others are removed. This was one of the few solutions that was kind of semi-automatic and not completely manual.

**Reply by mutambaraf** [reply]
@bolmsten I think the feedback we got from our user office was to avoid assigning multiple instruments but allow instrument scientist to assign instrument as the proposal goes through the process.

**Reply by bolmsten** [reply]
The question is how will we make it so that the instrument scientist see the proposals and also we need to open up for them to do proposal assignment to instrument

I can discuss this with our user officer and maybe they can make a suggestion with your.

**Reply by simonfernandes** [reply]
We need to clarify some things, but there is a possibility that the users on the instruments that are attached to the technique are not the correct people that we want to give this permission to.

If the list of instrument users is fine, Fredrik's suggestion to assign all instruments initially could work, but the team and Emma found it counter-intuitive to have to remove instruments, and we were concerned it might confuse scientists, so we weren't so keen on this.

If we can't use the instrument users, we may need to assign users to the technique and provide permissions that way. But we then need to consider whether users who are not listed in the technique list of users but are in the assigned instrument's users should have permissions or not.

A problem with all of the approaches involving getting permissions from assigned instruments may also be that we want everyone on the technique (i.e. everyone across all instruments assigned to the technique) to be able to see the proposal, even once a single instrument is chosen for the proposal.

Emma was open to the idea of initially assigning one (somewhat random) instrument from the technique (it's preferable to all instruments), but she hoped we wouldn't have to do this, and maybe it's too arbitrary for you at the ESS. If we assigned the correct technique users to every single instrument in the technique, then I think this method would be feasible.

So I think we could really do with getting more detail on how we want permissions to work on our side to help with this decision (we have an ongoing discussion in https://github.com/UserOfficeProject/issue-tracker/issues/1079). We have a meeting with Emma next Friday, but she can't answer all of these questions and so we will likely need to consult other stakeholders too.

**Reply by bolmsten** [reply]
My suggestion was to create a proposal status that automatically would remove the instruments that did not have a technical review. As for instrument scientist seeing proposal not belonging to their instrument I would suggest to simply add them to the other instrument, I am thinking that if the overlap between different MX beamlines is that big the scientist could belong to both.

**Reply by simonfernandes** [reply]
> My suggestion was to create a proposal status that automatically would remove the instruments that did not have a technical review.

We kept this in mind, and having now done some playing around with this it does seem to be more logical than initially thought. We have a meeting with Emma on Friday now and we'll see what she thinks of assigning all instruments and the workflow in general.

>  As for instrument scientist seeing proposal not belonging to their instrument I would suggest to simply add them to the other instrument, I am thinking that if the overlap between different MX beamlines is that big the scientist could belong to both.

I think this would mean that we'd need to create separate (duplicate) instruments, because we want to maintain how instruments work in a non-Xpress context. We'll have to run this by Emma again and it will depend on how we need to configure permissions.

--- 

We have also made a flow diagram to better help us and Emma understand this workflow and make sure we're on the same page. Please feel to comment or correct anything:

<img src="https://github.com/user-attachments/assets/500ae186-c118-4089-a811-16520d9ff0ef" width="200" height="400" />

**Reply by simonfernandes** [reply]
Hi @bolmsten. We've discussed this with Emma now and we unfortunately we don't think this workflow is going to be suitable for us due to a number of complications including:

- Scientists need to be able to change the status of a proposal to say that the proposal is under review and they have started looking at it. This has to be a manual step rather than triggered through a workflow. This is separate from the review. And at a future date, the scientist would change the status again once the experiment has run.
- Ideally we don't want to have to create separate Xpress instruments
- There's a lot of info and features that we want to hide on the main management page, and things we'd like to add which would make development challenging
- Similarly there are also things in the technical review that would need changing.
- We are limited in how permissions work, and any changes down the line become difficult. Instruments can work in different ways, e.g. sometimes not everyone should get notified.

We are keen to use a separate page hidden behind a feature flag and add users to the technique. It would provide us with a lot of flexibility in implementing permissions and other quirks of the Xpress system, and make it much easier for our scientists to use by providing them with a page for their proposals and a simple drop-down of statuses.

What are your thoughts on this?

**Reply by bolmsten** [reply]
Hi Simon,

I just spoke with Emma and maybe we are converging on a solution. 

If we can avoid it we would like to not have a new role in the system, instead continue using the instrument role. 

After listening to Emma, my suggestion would be to make instruments into instruments/techniques. When you create a new instrument/technique you also indicate via checkbox if it is a instrument or not. You then attach the instruments/techniques for the call and the user selects via the instrument/technique picker the one it wants. Then follows the normal flow with feasibility review etc ( with the feasibility review made dynamic (ELI working on)). You also open up a new page for the instrument scientist for proposals that has in the proposal workflow "enhanced feasiblity review", this page would allow them to change status and instrument.  When ready the proposal can be assigned the final instrument and the feasibility review is transferred. 

This solution would mean that we re-use the role and the review workflow. 

/Fredrik

**Reply by mutambaraf** [reply]
This means we have to allow instruments/techniques to attach to another instruments/techniques. I also think keeping techniques  and instruments as separate entities is a good idea given they are completely different and from the requirements techniques looks like a processes using a given instrument(s) , while instruments are physical objects. In addition we will still want to streamline the review process to be as simple as possible so we may want to avoid using feasibility review but use simple table. @simonfernandes

**Reply by simonfernandes** [reply]
Hi @bolmsten, apologies for the late update, I'm just back from leave. Here is what we decided:

- Keep a concept of techniques separate from instruments
  - Techniques have instruments attached to them
- Use a technique picker
  - Proposals are assigned to techniques on save/submission via the picker
- Use a separate Xpress management page (proposal table) hidden behind a feature flag, accessible to user officers and instrument scientists
  - Show extra information in the table via new columns, and hide many of the existing columns
  - Scientists can assign instruments to the proposal, limited by technique chosen
  - Scientists can change the status of a proposal (but only display a subset to choose from)
  - Scientists can filter proposals by date (possibly with Xpress-specific logic)
  - Scientists can export a spreadsheet (possibly with Xpress-specific logic)
  - Scientists can easily add comments to the proposal (possibly in-line in a table column)
- Add a call option to only show Xpress proposals in the Xpress management page

**Reply by bashanlam** [reply]
Proposed detail for 'Create an option to specify Xpress calls'
https://github.com/UserOfficeProject/issue-tracker/issues/1139

**Reply by bolmsten** [reply]
Feedback from ESS:

We think that we could leverage the work ELI has done but making it for technical review. For a XPRESS call you create a technical review template that allows for instrument and status action change by the instrument scientist. Adding new filters to the main proposal table should be possible and a new export. 

If we need to pre-select columns this could be when setting up the call.
